lib/integrations/awsoidc/eks_enroll_clusters.go

```diff
@@ -0,0 +1,472 @@
+	DeleteAccessEntry(ctx context.Context, params *eks.DeleteAccessEntryInput, optFns ...func(*eks.Options)) (*eks.DeleteAccessEntryOutput, error)
+
+	// DescribeCluster returns detailed information about an EKS cluster.
+	DescribeCluster(ctx context.Context, params *eks.DescribeClusterInput, optFns ...func(*eks.Options)) (*eks.DescribeClusterOutput, error)
+
+	// GetCallerIdentity returns details about the IAM user or role whose credentials are used to call the operation.
+	GetCallerIdentity(ctx context.Context, params *sts.GetCallerIdentityInput, optFns ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error)
+
+	// CheckAgentAlreadyInstalled checks if teleport-kube-agent Helm chart is already installed on the EKS cluster.
+	CheckAgentAlreadyInstalled(genericclioptions.RESTClientGetter, logrus.FieldLogger) (bool, error)
+
+	// InstallKubeAgent installs teleport-kube-agent Helm chart to the EKS cluster.
+	InstallKubeAgent(context.Context, *eksTypes.Cluster, string, genericclioptions.RESTClientGetter, logrus.FieldLogger, EnrollEKSClustersRequest) error
+}
+
+type defaultEnrollEKSClustersClient struct {
+	*eks.Client
+	stsClient *sts.Client
+}
+
+func (d *defaultEnrollEKSClustersClient) GetCallerIdentity(ctx context.Context, params *sts.GetCallerIdentityInput, optFns ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error) {
+	return d.stsClient.GetCallerIdentity(ctx, params, optFns...)
+}
+
+func (d *defaultEnrollEKSClustersClient) CheckAgentAlreadyInstalled(kubeconfig genericclioptions.RESTClientGetter, log logrus.FieldLogger) (bool, error) {
+	actionConfig, err := getHelmActionConfig(kubeconfig, log)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	return checkAgentAlreadyInstalled(actionConfig)
+}
+
+func (d *defaultEnrollEKSClustersClient) InstallKubeAgent(ctx context.Context, eksCluster *eksTypes.Cluster, proxyAddr string, kubeconfig genericclioptions.RESTClientGetter, log logrus.FieldLogger, req EnrollEKSClustersRequest) error {
+	actionConfig, err := getHelmActionConfig(kubeconfig, log)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return installKubeAgent(ctx, eksCluster, proxyAddr, actionConfig, getHelmSettings(), req)
+}
+
+// NewEnrollEKSClustersClient returns new client that can be used to enroll EKS clusters into Teleport.
+func NewEnrollEKSClustersClient(ctx context.Context, req *AWSClientRequest) (EnrollEKSCLusterClient, error) {
+	eksClient, err := newEKSClient(ctx, req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	stsClient, err := newSTSClient(ctx, req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clt := defaultEnrollEKSClustersClient{
+		Client:    eksClient,
+		stsClient: stsClient,
+	}
+
+	return &clt, nil
+}
+
+// EnrollEKSClustersRequest contains the required fields to enroll EKS cluster to Teleport.
+type EnrollEKSClustersRequest struct {
+	// Region is the AWS Region.
+	Region string
+
+	// ClusterNames is name of the EKS cluster to enroll.
+	ClusterNames []string
```

Comment:
Given that we accept multiple cluster names, should we accept one resource id per cluster name so we can trace which agent is not successfully installed?

Reply:
Refactored it, so now tokens are generated on the backend and we return resourceId per cluster so UI can wait on them correctly - https://github.com/gravitational/teleport/pull/36594/commits/fb4438c01624a417bdbb346c0624ea13d531582f